### PR TITLE
chore: standardise how unit/integration tests is run on CI, adding a build command to compress the system-tests files

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
This PR adds two things:

- a `mage packageSystemTests` command, plugged into filebeat and metricbeat modules (both OSS and xpack flavours).
- a `Tar` method under the `commons.go` mage build file, so that it's possible to compress a source directory using TAR+GZIP algorithms, and writing it to the build dir in the root directory of the Beats repository.

Besides that, we are removing a python script that detected the path to the Python system-tests (used in the pipeline to store the folder name in a variable), and we are updating the Jenkins pipeline to call the new build system command instead of the previous script. To keep existing behaviour, that we could consider it wrong and would possible discusse about it in a follow-up PR, we are passing a new argument to the groovy method in the pipeline, representing the directory in which the Beat live (i.e. filebeat, x-pack/filebeat), so that it's possible to run the `mage` command in the right context.

Finally, we are keeping Pipeline2.0-consistency across Beats calling the Python system-tests, using OSS metricbeat's as the canonical reference: run unit tests, then Go integration tests, and finally Python integration tests when needed. This change applies to filebeat and metricbeat in both OSS and xpack flavours.

**UPDATE**: I finally revisited each Beat and made sure all of them run the unit, goInteg and pythonInteg tests in a consistent manner: instead of calling the obscure mage test command, we separate it in three different stages when needed: `mage unitTest`, `mage goIntegTest` and `mage pythonIntegTest`. I've verified one by one that each of those three commands is present in the build for the related Beat, adding or removing them from the CI descriptor (`Jenkinsfile.yml`) when/if needed, creating the consistent experience of running atomic commands for each test type.

## Why is it important?
First one is simplicity: instead of adding code into the Jenkins pipeline that makes it almost impossible to reproduce it locally, we are moving to the build system, which will allow developers to run this target locally and verify what the CI will do.

Second is consistency: metricbeat and filebeat are running system-tests, and we want to operate them on CI in the same manner. That's why we are making the descriptors very similar.

And third, separation of concers: we want to run unit > integration > e2e tests in different stages, so that we acknowledge the power of not building/running code that could have been broken after a possible failure in a previous stage.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ] 

## How to test this PR locally

```shell
$ cd filebeat
$ mage pythonIntegTest
$ mage packageSystemTests
>> skipping symlink: build/system-tests/run/test_harvester.Test.test_ignore_symlink/log/symlink.log
>> skipping symlink: build/system-tests/run/test_harvester.Test.test_symlink_and_file/log/symlink.log
>> skipping symlink: build/system-tests/run/test_harvester.Test.test_symlink_rotated/log/symlink.log
>> skipping symlink: build/system-tests/run/test_harvester.Test.test_symlinks_enabled/log/symlink.log
>> skipping symlink: build/system-tests/run/test_harvester.Test.test_truncate/log/symlink.log
>> skipping symlink: build/system-tests/run/test_input.Test.test_skip_symlinks/log/test.log

# checking file system contents
$ find build/system-tests/run -name "*" | wc -l
    3940
    
# checking TAR contents
$ tar -tf ../build/system-tests-filebeat.tar.gz | wc -l
    3934
```

**Expected behaviour:** a TAR file in the root's `build` directory named as the Beat will exist (i.e. ./build/system-tests-filebeat.tar.gz)

## Related issues

- Relates #26208

## Follow-ups
We should consider following same separation of concerns in all Beats running `mage build test`, so that they run unitTest instead, and then having separate stages for integrations tests.

### CODE REVIEW

File: dev-tools/mage/pkg.go
Comment (mdelapenya) [inline dev-tools/mage/pkg.go:148]:
```suggestion
// PackageSystemTests pacakges the python system tests results
```

File: x-pack/filebeat/Jenkinsfile.yml
Comment (v1v) [inline x-pack/filebeat/Jenkinsfile.yml:40]: I think we don't need this particular flag, this is only required for the ITs to identify what module has been changed and run only for that particular module if possible.

File: x-pack/filebeat/Jenkinsfile.yml
Comment (v1v) [inline x-pack/filebeat/Jenkinsfile.yml:44]: I'm not sure if the withModule is required for the Go ITs

File: dev-tools/mage/common.go
Comment (kuisathaverat) [inline dev-tools/mage/common.go:356]: Could it be a CLI separate tool?

Comment (mdelapenya) [reply dev-tools/mage/common.go:356]: It could be running `tar` directly! 😂 But I wanted to embed it into the build system for cross-platform compatibility

File: x-pack/filebeat/Jenkinsfile.yml
Comment (v1v) [reply x-pack/filebeat/Jenkinsfile.yml:44]: It's also used in the goIntegTest -> https://github.com/elastic/beats/pull/19908#issuecomment-658927205 and https://github.com/elastic/beats/pull/18592#issuecomment-633618743

So I said nothing regarding the `goIntegTest` stage

File: x-pack/filebeat/Jenkinsfile.yml
Comment (mdelapenya) [reply x-pack/filebeat/Jenkinsfile.yml:40]: You are right, I must have copied an entire block wrong. Will send it ASAP

**Review by v1v — APPROVED** [review]
💯

File: dev-tools/mage/common.go
Comment (jsoriano) [inline dev-tools/mage/common.go:422]: Consider using [`ioutil.WriteFile`](https://golang.org/pkg/io/ioutil/#WriteFile) for this.
```suggestion
	return ioutil.WriteFile(targetFile, buf.Bytes(), os.FileMode(0600))
```

File: dev-tools/mage/common.go
Comment (jsoriano) [inline dev-tools/mage/common.go:395]: This file should be closed to avoid keeping too many open files.
```suggestion
			data, err := os.Open(file)
			if err != nil {
				fmt.Printf(">> error opening file: %s\n", err)
				return err
			}
			defer data.Close()
```

File: dev-tools/mage/pkg.go
Comment (jsoriano) [inline dev-tools/mage/pkg.go:148]: Umm `dev-tools/mage/pkg.go` contains targets and helpers to package builds of beats. I'd say this new target is more related to tests, consider moving it to other file.

You can also add it to to some directory under `dev-tools/mage/target` so it can also be imported in magefiles.

**Review by jsoriano — COMMENTED** [review]
:+1: added some suggestions.

File: dev-tools/mage/pkg.go
Comment (mdelapenya) [reply dev-tools/mage/pkg.go:148]: Ok, I'll move this helper method to another file. Thanks for the review!

File: dev-tools/mage/common.go
Comment (kuisathaverat) [reply dev-tools/mage/common.go:356]: cross-platform is super nice, this is something we were talking @v1v and me the other day, this, for example, is one of those cases that we need to do something on all platforms and there is no common way to do that, on each platform we have to install and use similar tools but not the same. To have a goTar CLI tool that only needs Go (directly install the binary from the platform kit from a bucket or something) will simplify our pipelines and build systems.

**Review by cachedout — APPROVED** [review]

File: x-pack/elastic-agent/magefile.go
Comment (mdelapenya) [inline x-pack/elastic-agent/magefile.go:35]: @jsoriano is it possible to include the command once, and apply it to all subprojects/beats?

Comment (jsoriano) [reply x-pack/elastic-agent/magefile.go:35]: Not that I know, but maybe an idea is to add the target to `dev-tools/mage/target/integtest` (instead of to `dev-tools/mage/target/systemtest`), so it will be included in all projects that have go or python integration tests and already include these targets.

Comment (mdelapenya) [reply x-pack/elastic-agent/magefile.go:35]: I like the idea: will add a commit on top, thanks!

File: dev-tools/mage/common.go
Comment (mdelapenya) [reply dev-tools/mage/common.go:356]: I really like the idea of extracting "things" / commands out to tools. You got me with the unit tests aspect :) 

@jsoriano, wdyt? Can we move on this, or create it a separate tool.

Comment (jsoriano) [reply dev-tools/mage/common.go:356]: Having a set of common CLI tools that work exactly the same in different platforms would be definitely awesome, in many contexts :slightly_smiling_face:

But I also see some potential problems. These tools would become an additional dependency that would need to be installed in CI and in all developers laptops, for newcomers this can suppose an additional thing to handle and learn. Of course this can be overcome by adding some helpers to build tools that transparently install and use the tools, but then this would be another piece to maintain in any new build tool, and maintainers of build scripts would still have the cognitive load of remembering to use these CLI tools instead of the common thing in their platform or language.

If the overhead of supporting this on every build tool is significatively smaller than the overhead of implementing tar and some other common commands in every project where this is needed, then great, but I would be a bit cautious by now. It may happen that [this divergence](https://en.wikipedia.org/wiki/Tar_(computing)#Key_implementations) of behaviours happens in tar but not in so many other common tools.

In summary, I would postpone this for a future and broader conversation.

Comment (mdelapenya) [reply dev-tools/mage/common.go:356]: Cool, I think we can merge this one then, and maybe use it for a future follow-up discussion being the case we need to implement those specific tools.

**Comment by mdelapenya** [comment]
Jenkins run the tests please

File: Jenkinsfile
Comment (v1v) [inline Jenkinsfile:781]: Can it be done this change in a different PR?  Then, we can remove the scripts that configure the tools too?

File: dev-tools/mage/target/integtest/notests/integtest.go
Comment (jsoriano) [inline dev-tools/mage/target/integtest/notests/integtest.go:22]: I think this line is not needed here.

File: Jenkinsfile
Comment (cachedout) [inline Jenkinsfile:782]: What happens if this is `''` as it would be if the `.get()` returned its default value?

File: dev-tools/mage/target/integtest/integtest.go
Comment (jsoriano) [inline dev-tools/mage/target/integtest/integtest.go:102]: Nit. This is a bit cryptic, could `filepath.Join` be used instead?
```suggestion
	systemTestsDir := filepath.Join("build", "system-tests", "run")
```

File: Jenkinsfile
Comment (mdelapenya) [reply Jenkinsfile:785]: We are setting the variable with `setEnvVar('GO_VERSION', readFile(".go-version").trim())`. Do you mean that the `readFile` can return empty value? That would be the case if the `.go-version` file disappears, which will represent a breaking change in how Go is installed.

**Comment by mdelapenya** [comment]
jenkins run the tests please

**Comment by mdelapenya** [comment]
jenkins run the tests please

**Comment by mdelapenya** [comment]
@jsoriano I think this is ready to merge. Waiting for final 👍 to do so

File: heartbeat/Jenkinsfile.yml
Comment (jsoriano) [inline heartbeat/Jenkinsfile.yml:47]: Same here, not sure if we can use `withModule`. Heartbeat doesn't have modules, it has monitors, but they have a different directory layout.

File: libbeat/Jenkinsfile.yml
Comment (jsoriano) [inline libbeat/Jenkinsfile.yml:39]:
```suggestion
```

File: x-pack/functionbeat/Jenkinsfile.yml
Comment (jsoriano) [inline x-pack/functionbeat/Jenkinsfile.yml:38]:
```suggestion
```

File: x-pack/heartbeat/Jenkinsfile.yml
Comment (jsoriano) [inline x-pack/heartbeat/Jenkinsfile.yml:29]:
```suggestion
```

File: x-pack/heartbeat/Jenkinsfile.yml
Comment (jsoriano) [inline x-pack/heartbeat/Jenkinsfile.yml:27]: I think heartbeat doesn't have integration tests.

File: x-pack/libbeat/Jenkinsfile.yml
Comment (jsoriano) [inline x-pack/libbeat/Jenkinsfile.yml:45]:
```suggestion
        stage: mandatory
    pythonIntegTest:
        mage: "mage pythonIntegTest"
```

File: x-pack/osquerybeat/Jenkinsfile.yml
Comment (jsoriano) [inline x-pack/osquerybeat/Jenkinsfile.yml:27]:
```suggestion
```

File: Jenkinsfile
Comment (jsoriano) [inline Jenkinsfile:797]: If `packageSystemTests` is only called here, are all the other changes required? I am a bit concerned that we may be missing the execution of some tests in some cases, it may be good to separate these changes to a different PR if possible, so we can merge this feature with less risks.

**Review by jsoriano — CHANGES_REQUESTED** [review]
Requesting changes mainly for the use of `withModule`, I am not sure if this can be generaly used in other Beats beyond Metricbeat.

I am also a bit worried about the refactor of mage targets, we may be missing the execution of some tests by mistake. I think that the use of a central `test` target was introduced for this. I wouldn't be in any case opposed to the use more specific targets, specially now that we have separated Jenkinsfiles, but I would like to have a second opinion by @andrewkroh about this.

File: Jenkinsfile
Comment (mdelapenya) [reply Jenkinsfile:797]: Ok, let me split the PR in two:
1. changing test execution layout on CI
2. adding the new mage command

**Comment by mdelapenya** [comment]
Ok, this PR is not needed anymore, as it has been superseded by https://github.com/elastic/beats/pull/27069 and https://github.com/elastic/beats/pull/27295

**Pull request closed without merging**
